src/mongoc/mongoc-database.c

```diff
@@ -618,6 +618,7 @@ mongoc_database_has_collection (mongoc_database_t *database,
        (error &&
         ((error->domain != 0) ||
          (error->code != 0)))) {
+      mongoc_cursor_destroy( cursor );
```

Comment:
You are of course right, I had only tested it with an error, not an invalid cursor. I will fix this tonight by separating the if clause for cursor and error.